src/distributions/reusable_weighted_index.rs

```diff
@@ -0,0 +1,335 @@
+/// supports `O(1)` sampling, but with much higher initialisation cost.
+///
+/// A `ReusableWeightedIndex<X>` contains a `Vec<X>` and a [`Uniform<X>`] and so its
+/// size is the sum of the size of those objects, possibly plus some alignment.
+///
+/// Creating a `ReusableWeightedIndex<X>` will allocate enough space to hold `N - 1`
+/// weights of type `X`, where `N` is the number of weights. However, since
+/// `Vec` doesn't guarantee a particular growth strategy, additional memory
+/// might be allocated but not used. Since the `ReusableWeightedIndex` object also
+/// contains, this might cause additional allocations, though for primitive
+/// types, [`Uniform<X>`] doesn't allocate any memory.
+///
+/// Reusing `ReusableWeightedIndex<X>` does not cause additional memory allocations as
+/// long as the `ReusableWeightedIndex` is not dropped and number of weights is not
+/// increased.
+///
+/// # Panics
+///
+/// Panics if the sum of the weights is zero.
+///
+/// # Implementation
+///
+/// The `ReusableWeightedIndex` object contains a `Vec<X>` and a [`Uniform<X>`].
+/// When sampling from the distribution, the `Uniform` object is used to sample
+/// from the `Vec`. The `Vec` is reused, so it is not necessary to drop the
+/// `ReusableWeightedIndex before creating a new one.
+///
+/// Sampling from `ReusableWeightedIndex` will result in a single call to
+/// `Uniform<X>::sample` (method of the [`Distribution`] trait), which typically
+/// will request a single value from the underlying [`RngCore`], though the
+/// exact number depends on the implementation of `Uniform<X>::sample`.
+///
+/// # Example
+///
+/// ```
+/// use rand::prelude::*;
+/// use rand::distributions::CumulativeWeightsWrapper;
+///
+/// let choices = ['a', 'b', 'c'];
+/// let weights = [2,   1,   1];
+/// let mut dist = CumulativeWeightsWrapper::new();
+/// let dist = dist.fill(&weights).unwrap();
+/// let mut rng = thread_rng();
+/// for _ in 0..100 {
+///     // 50% chance to print 'a', 25% chance to print 'b', 25% chance to print 'c'
+///     println!("{}", choices[dist.sample(&mut rng)]);
+/// }
+///
+/// let items = [('a', 0), ('b', 3), ('c', 7)];
+/// let mut dist2 = CumulativeWeightsWrapper::new();
+/// let dist2 = dist2.fill(items.iter().map(|item| item.1)).unwrap();
+/// for _ in 0..100 {
+///     // 0% chance to print 'a', 30% chance to print 'b', 70% chance to print 'c'
+///     println!("{}", items[dist2.sample(&mut rng)].0);
+/// }
+/// ```
+///
+/// [`Uniform<X>`]: crate::distributions::Uniform
+/// [`RngCore`]: crate::RngCore
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct ReusableWeightedIndex<'a, X: SampleUniform + PartialOrd> {
+    wrapper: &'a CumulativeWeightsWrapper<X>,
+    total_weight: X,
+    weight_distribution: X::Sampler,
+}
+
+/// Just a wrapper to surround internal persistent storage to reuse in [`ReusableWeightedIndex`]
+/// instances.
+#[derive(Debug, Clone, PartialEq)]
+pub struct CumulativeWeightsWrapper<X> {
```

Comment:
May be it will be useful to replace it with slice container to allow somebody to use stack-allocated container (via smallvec for example)